guillotina/auth/utils.py

```diff
@@ -0,0 +1,66 @@
+from datetime import datetime
+from datetime import timedelta
+from typing import Optional
+
+import jwt
+from guillotina import task_vars
+from guillotina._settings import app_settings
+from guillotina.auth import groups  # noqa
+from guillotina.auth.users import ROOT_USER_ID
+from guillotina.component import get_utility
+from guillotina.interfaces import IApplication
+from guillotina.interfaces import IPrincipal
+from guillotina.interfaces import ISecurityPolicy
+from guillotina.profile import profilable
+
+
+@profilable
+async def authenticate_request(request) -> Optional[IPrincipal]:
+    for policy in app_settings['auth_extractors']:
+        token = await policy(request).extract_token()
+        if token:
+            for validator in app_settings['auth_token_validators']:
+                if (validator.for_validators is not None and
+                        policy.name not in validator.for_validators):
+                    continue
+                user = await validator().validate(token)
+                if user is not None:
+                    set_authenticated_user(user)
+                    return user
+    set_authenticated_user(None)
+    return None
+
+
+def set_authenticated_user(user):
+    policy = get_utility(ISecurityPolicy)
+    policy.invalidate_cache()
+    if user is not None:
+        if hasattr(user, 'roles') and 'guillotina.Authenticated' not in user.roles:
+            user.roles['guillotina.Authenticated'] = 1
+    task_vars.authenticated_user.set(user)
+    task_vars.authenticated_user_groups.set({})
```

Comment:
I think that groups should be a parameter of the user

Reply:
This is now cache on principal instead.